radix-engine-queries/src/typed_native_events.rs

```diff
@@ -0,0 +1,647 @@
+//! the toolkit and consumed by the gateway for some of its internal operations.
+
+use crate::typed_substate_layout::*;
+use radix_engine::blueprints::native_schema::*;
+use radix_engine::blueprints::pool::multi_resource_pool::*;
+use radix_engine::blueprints::pool::one_resource_pool::*;
+use radix_engine::blueprints::pool::two_resource_pool::*;
+use radix_engine::types::*;
+use radix_engine_interface::api::node_modules::auth::*;
+use radix_engine_interface::api::*;
+use radix_engine_interface::blueprints::access_controller::*;
+use radix_engine_interface::blueprints::account::*;
+use radix_engine_interface::blueprints::consensus_manager::*;
+use radix_engine_interface::blueprints::identity::*;
+
+pub fn to_typed_native_event(
+    event_type_identifier: &EventTypeIdentifier,
+    event_data: &[u8],
+) -> Result<TypedNativeEvent, TypedNativeEventError> {
+    let typed_native_event_key =
+        resolve_typed_event_key_from_event_type_identifier(event_type_identifier)?;
+    to_typed_event_with_event_key(&typed_native_event_key, event_data)
+}
+
+fn resolve_typed_event_key_from_event_type_identifier(
+    event_type_identifier: &EventTypeIdentifier,
+) -> Result<TypedNativeEventKey, TypedNativeEventError> {
+    let local_type_index = match event_type_identifier.1 {
+        TypePointer::Package(_, x) => x,
+        _ => panic!(""),
+    };
+
+    match &event_type_identifier.0 {
+        /* Method or Function emitter on a known node module */
+        Emitter::Method(_, ObjectModuleId::AccessRules)
+        | Emitter::Function(_, ObjectModuleId::AccessRules, ..) => {
+            TypedAccessRulesBlueprintEventKey::new(
+                &ACCESS_RULES_PACKAGE_DEFINITION,
+                ACCESS_RULES_BLUEPRINT,
+                &local_type_index,
+            )
+            .map(TypedNativeEventKey::from)
+        }
+        Emitter::Method(_, ObjectModuleId::Metadata)
+        | Emitter::Function(_, ObjectModuleId::Metadata, ..) => {
+            TypedMetadataBlueprintEventKey::new(
+                &METADATA_PACKAGE_DEFINITION,
+                METADATA_BLUEPRINT,
+                &local_type_index,
+            )
+            .map(TypedNativeEventKey::from)
+        }
+        Emitter::Method(_, ObjectModuleId::Royalty)
+        | Emitter::Function(_, ObjectModuleId::Royalty, ..) => {
+            TypedComponentRoyaltyBlueprintEventKey::new(
+                &ROYALTY_PACKAGE_DEFINITION,
+                COMPONENT_ROYALTY_BLUEPRINT,
+                &local_type_index,
+            )
+            .map(TypedNativeEventKey::from)
+        }
+
+        /* Functions on well-known packages */
+        Emitter::Function(node_id, ObjectModuleId::Main, blueprint_name) => {
+            let package_address = PackageAddress::try_from(node_id.as_bytes())
+                .expect("Function emitter's NodeId is not a valid package address!");
+
+            match package_address {
+                PACKAGE_PACKAGE => TypedPackagePackageEventKey::new(
```

Comment:
> Also, we're repeating TypedPackagePackageEventKey::new(&x, &y, &z).map(TypedNativeEventKey::from) a lot - could we instead create a function eg to_native_event_key(definition, blueprint_name, local_type_index)?

I don't think that we have enough information to construct a `TypedNativeEventKey` based on the definition, blueprint name, and local type index alone, we need some added information on what package this blueprint belongs to so that we can do the mapping. If we add an extra parameter of package identifier/name, we end up with a large match statement in `to_native_event_key` which might just be redundant.

Reply:
Right - I think most of this is handled by my suggestion for the `match package_address => Option<PackageDefinition>` and then create the `TypedNativeEventKey` in one place from the `Option<PackageDefinition>`